src/main.rs

```diff
@@ -27,37 +36,76 @@ fn main() {
     };
 
     let local_dir_absolute_path = match env::current_dir() {
-        Err(_) => exit_with_error(&"Could not resolve working directory, make sure it exists and user has enough permissions to work with it.", 1),
+        Err(_) => exit_with_error("Could not resolve working directory, make sure it exists and user has enough permissions to work with it.", 1),
         Ok(value) => fs::canonicalize(value).unwrap()
     };
 
-    let mut config_file = local_dir_absolute_path.to_owned();
+    let mut config_file = local_dir_absolute_path.clone();
     config_file.push(".mainframer/config.yml");
 
     let config = match merge_configs(&config_file) {
         Err(error) => exit_with_error(&error, 1),
         Ok(value) => value
     };
 
-    let ignore = Ignore::from_working_dir(&local_dir_absolute_path);
+    let ignore = Ignore::from_working_dir(&local_dir_absolute_path.clone());
 
-    let start = Instant::now();
+    println!("Pushing...");
 
-    if let Err(error) = sync_before_remote_command(&local_dir_absolute_path, &config, &ignore) {
-        exit_with_error(&format!("Sync local → remote machine failed: {}.", error), 1)
+    match sync::push(&local_dir_absolute_path, &config, &ignore) {
+        Err(err) => exit_with_error(&format!("Push failed: {}, took {}", err.message, format_duration(err.duration)), 1),
+        Ok(ok) => println!("Push done: took {}.\n", format_duration(ok.duration)),
     }
 
-    let remote_command_result = execute_remote_command(&local_dir_absolute_path, &args, &config);
+    match config.pull.mode {
+        PullMode::Serial => println!("Executing command on remote machine...\n"),
+        PullMode::Parallel(_) => println!("Executing command on remote machine (pulling in parallel)...\n")
+    }
+
+    let mut remote_command_readers = remote_command::execute_remote_command(
+        args.command.clone(),
+        config.clone(),
+        sync::project_dir_on_remote_machine(&local_dir_absolute_path.clone()),
+        2
+    );
+
+    let pull_finished_rx = sync::pull(&local_dir_absolute_path, config.clone(), ignore, &config.pull.mode, remote_command_readers.pop().unwrap());
 
-    if let Err(error) = sync_after_remote_command(&local_dir_absolute_path, &config, &ignore) {
-        exit_with_error(&format!("Sync remote → local machine failed: {}.", error), 1)
+    let remote_command_result = remote_command_readers
+        .pop()
+        .unwrap()
+        .recv()
+        .unwrap();
+
+    match remote_command_result {
+        Err(ref err) => eprintln!("\nExecution failed: took {}.\nPulling...", format_duration(err.duration)),
```

Comment:
The `Pulling...` part of this message seems a bit misleading in the parallel case, as I'd expect one last incremental sync. Consider having separate messaging for the serial vs parallel cases.

Reply:
While it's incremental, it's nontheless *potentially* long

In many cases remote command produces file(s) at the very end, thus parallel pulling might not be parallel